segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -639,10 +640,9 @@ private int readInternal(SegmentHandle handle, byte[] buffer, long offset, int b
         //There is only one file per segment.
         FileStatus currentFile = findStatusForSegment(handle.getSegmentName(), true);
         try (FSDataInputStream stream = this.fileSystem.open(currentFile.getPath())) {
-            if (offset + length > stream.available()) {
-                throw new ArrayIndexOutOfBoundsException();
-            }
             stream.readFully(offset, buffer, bufferOffset, length);
+        } catch (EOFException e) {
+            throw new IllegalArgumentException(String.format("Reading at offset (%d) which is beyond the current size of segment.", offset));
```

Comment:
This is not what I asked...

Reply:
1. It does not make sense to just consume an error in upper layers and return lesser number of bytes.
2. This is what other tier-2 impls do. It makes sense to keep the behavior similar.